Make recents understand context

escherize opened the pull request:
Solves #43300 

Adds 2 endpoints, and deprecates 1.

Updates client to send a request to register selection events, and pull selection events in the entity picker.

Doesn't log recent views from pinned collection card-reads.

escherize committed:
adds the endpoint and a test

escherize committed:
add recents endpoint, todo: tests

escherize committed:
add post recents endpoint

escherize committed:
return recent views for both lists, but make the endpoint work

escherize committed:
Make recent-views context aware

- pruning is context aware, only checks for the recently-inserted
  context
- Adds endpoints:
    - POST to create selection recents,
    - GET activity/recents
      - requres context query param to be one of:
      - all, views, selections

- Adds context arg to update-users-recent-views!
- Cleans up arg schema for update-users-recent-views

escherize committed:
impl GET api/activity/recents

- return recent-selections and recent-views from
- send context == collection from pinned card reads

escherize committed:
Merge branch 'master' into recents-with-context

escherize committed:
update callsites of recent-views/update-users-recent-views!

- to use :view param where necessary

escherize committed:
fixes models/recent-view tests

escherize committed:
adds more activity/recent-view tests

escherize committed:
wip

- fix whitespace linter

escherize committed:
Merge remote-tracking branch 'origin' into recents-with-context

escherize committed:
Fix command palette e2e test

- reuse util snake-keys

sloansparger committed:
updates fe to use new recents endpoints

sloansparger committed:
fixes fe type issue

escherize committed:
snake-keys -> deep-snake-keys

- I've been betrayed by lsp rename

escherize committed:
snake-keys -> deep-snake-keys

- I've been betrayed by lsp rename

escherize committed:
log selection events for created collections

escherize committed:
mysql doesn't allow default values for TEXT types

escherize committed:
log a recent view on data-picker selection

escherize committed:
decouple view-log context from card-event view context

escherize commented on src/metabase/api/activity.clj:
```diff
@@ -112,11 +114,34 @@
 
-(api/defendpoint GET "/recent_views"
+(api/defendpoint ^:deprecated GET "/recent_views"
   "Get a list of 100 models (cards, models, tables, dashboards, and collections) that the current user has been viewing most
   recently. Return a maximum of 20 model of each, if they've looked at at least 20."
   []
-  {:recent_views (recent-views/get-list *current-user-id*)})
+  (u/snake-keys
+   (recent-views/get-recents *current-user-id* :views)))
```
We can basically get rid of this endpoint, it's functionally the same as `GET activity/recents?context=views`.

escherize commented on src/metabase/events/schema.clj:
```diff
@@ -37,7 +36,7 @@
 (let [view-only      (mc/schema
                       [:map {:closed true}
-                       [:context ::view-log/context]
+                       [:context [:enum :dashboard :question :collection]]
```
`::view-log/context` is not the right schema, since this event architecture is shared by more than view-log as a subscriber. (Also we started passing in `:collection`)

calherries replied:
Oops, the reason why I put `::view-log/context` here was to document the coupling of possible values of `view_log.context` to `:event/card-read`'s `context. But I didn't explain that anywhere or enforce it. I wanted to explicitly couple them because all card reads should theoretically logged in the `view_log` table now and into the future.

I've just created a PR to make sure `::view-log/context` doesn't get out of date again: [Actually validate view_log.context matches ::view-log/context on insert](https://github.com/metabase/metabase/pull/43801)

escherize commented on src/metabase/models/recent_views.clj:
```diff
@@ -69,16 +72,30 @@
    :dashboard :table :collection])
 
-(defn- ids-to-prune-for-user+model [user-id model]
+(defn- moi->type
```
I am purposely using a synthetic name (as opposed to a natural name) to refer to `models of interest` (here `moi`). We take them in as `model` on the activity api 🫠, and there's a mapping from `moi`s to models, but they aren't the same so I don't think we should call them the same thing.

noahmoss replied:
I can't help but read `moi` as a French word and pronounce it in my head accordingly

maybe `recent-views-model`/`rv-model` or something like that? to make it clear that it's the "model" field of recent views, rather than the normal toucan model identifier.

escherize replied:
(or the model card type).

Yea, I'll switch `moi` to another synthetic name.

escherize commented on src/metabase/models/recent_views.clj:
```diff
@@ -69,16 +72,30 @@
                              [:= :rv.user_id user-id]
-                             (when (#{:card :model} model)
+                             [:= :rv.context (h2x/literal (name context))]
+                             (when (#{:card :model} model) ;; TODO add metric
```
Some driveby work has been done to include metrics in our work, but I don't think it ever worked, and we do not query for them, so this should be addressed at some point in the future.

noahmoss replied:
Could you add a little more context in a comment or the ns docstring or something? I'm imagining viewing this ns in a year and not being able to figure out why these TODOs are here

escherize replied:
Good call, done.

escherize committed:
fix a doc typo

escherize committed:
Merge branch 'master' into recents-with-context

escherize committed:
stop double logging recent-views on POST

sloansparger committed:
maybe fixes some tests

sloansparger committed:
some e2e fixes

escherize committed:
Merge remote-tracking branch 'origin' into recents-with-context

escherize committed:
fix mysterious divide by zero during score search

escherize committed:
fix divide by zero possibilities everywhere in score-items

metabase.api.activity/score-items used to throw when there weren't any
items being scored (even though there's a `(when (seq items) ...)` check)

sloansparger committed:
more test fixes

noahmoss commented on src/metabase/api/activity.clj:
```diff
@@ -112,11 +114,33 @@
+  [:as {{:keys [model model_id context]} :body}]
+  {model (into [:enum] recent-views/models-of-interest)
+   model_id ms/PositiveInt
+   context [:enum :selection]}
```
Should this also include `:views` or is the docstring wrong?

escherize replied:
Docstring was wrong

noahmoss commented on src/metabase/models/recent_views.clj:
```diff
@@ -69,16 +72,30 @@
+    "card"
+    (name moi)))
+
+(mu/defn moi->model
+  "Given a model of interest, returns the model for it."
+  [moi :- (into [:enum] models-of-interest)]
+  (keyword "model" (str/capitalize (moi->type moi))))
```
take it or leave it but IMO this would be clearer as just a map like
```
{:model      :model/Card
 :card       :model/Card
 :dashboard  :model/Dashboard
 :table      :model/Table
 :collection :model/Collection}
```
it's a small fixed set so this makes the mapping super easy to understand instead of needing to think about multiple different functions

escherize replied:
Thanks! fixed

noahmoss commented on src/metabase/models/recent_views.clj:
```diff
@@ -69,16 +72,30 @@
 
-(defn- ids-to-prune-for-user+model [user-id model]
+(defn- moi->type
+  "Turns a model of interest into the base name of its table"
```
nit — it's not the name of the table, right? aside from `collection`, but the others are different

noahmoss commented on src/metabase/models/recent_views.clj:
```diff
@@ -55,8 +55,11 @@
+  [user-id context]
+  (->> (t2/select :model/RecentViews
+                  :user_id user-id
+                  :context (name context)
```
If you define a transformation on `:model/RecentViews` you can use this as a keyword directly in t2 calls

escherize replied:
fixed

escherize committed:
fix more e2e tests, + rename endpoint in tests

escherize committed:
Merge branch 'recents-with-context' of github.com:metabase/metabase into recents-with-context

noahmoss commented on src/metabase/models/recent_views.clj:
```diff
@@ -9,7 +9,7 @@
        see: [[update-users-recent-views!]]
```
The example above needs to be updated right? Maybe the docstring (or somewhere else in this namespace) could also describe what `context` is used for, particularly what `selection` means since I don't think that's a standard term elsewhere

escherize replied:
yes! fixed

noahmoss commented on src/metabase/util.cljc:
```diff
@@ -247,7 +247,12 @@
+  (update-keys m ->snake_case_en))
+
+(defn deep-snake-keys
+  "Convert the keys in a map to `snake_case`."
```
nit - probably obvious from the function name but `keys in a map` alone to me implies top-level 🤷‍♂️ 
```suggestion
  "Recursively convert the keys in a map to `snake_case`."
```

escherize replied:
Ah, good call. I copied the old docstring, but I like yours better.

escherize replied:
fixed

noahmoss reviewed (approved):
only reviewed the Clojure files, have some minor comments but mostly lgtm

sloansparger committed:
fix oopsie

sloansparger committed:
fixes a few more tests

escherize committed:
address review comments/questions

calherries commented on src/metabase/events/recent_views.clj:
```diff
@@ -42,41 +36,40 @@
+  [topic {:keys [object-id user-id context]}]
+  (when (= context :question)
+    ;; GET /api/card/:id will trigger a card-read event. We don't want to count _pinned collection views_ as recent-views,
+    ;; so a collection with pinned questions will send a request with context==:collection, which we skip below:
```
As of https://github.com/metabase/metabase/pull/43102 `:context` can be `:dashboard` too.
```
grep -r --include=\*.clj "event/card-read" src 

src/metabase/api/embed.clj:      (events/publish-event! :event/card-read {:object-id (:id <>), :user-id api/*current-user-id*, :context :question}))))
src/metabase/api/embed.clj:    (events/publish-event! :event/card-read {:object-id card-id, :user-id api/*current-user-id*, :context :dashboard})))
src/metabase/api/embed.clj:    (events/publish-event! :event/card-read {:object-id card-id, :user-id api/*current-user-id*, :context :dashboard})))
src/metabase/api/card.clj:        (events/publish-event! :event/card-read {:object-id (:id <>) :user-id api/*current-user-id* :context :question})))))
src/metabase/api/dashboard.clj:    (events/publish-event! :event/card-read {:object-id card-id, :user-id api/*current-user-id*, :context :dashboard})))
src/metabase/api/dashboard.clj:    (events/publish-event! :event/card-read {:object-id card-id, :user-id api/*current-user-id*, :context :dashboard})))
src/metabase/api/public.clj:    (events/publish-event! :event/card-read {:object-id (:id <>), :user-id api/*current-user-id*, :context :question})))
src/metabase/api/public.clj:      (events/publish-event! :event/card-read {:object-id card-id, :user-id api/*current-user-id*, :context :dashboard}))))
src/metabase/api/public.clj:      (events/publish-event! :event/card-read {:object-id card-id, :user-id api/*current-user-id*, :context :dashboard}))))
```

Should this comment also mention that card view with `context=dashboard` are also omitted from recent views? It implies that the only  purpose of this when clause is to exclude views where `context=collection`.

escherize replied:
Yeah I think updating comments so they're more informative totally worth it. I'll make the change.

escherize committed:
allow for a comma delimited list in qps like ?context=views,selections

returns all recent view items with those contexts, most recent first
under the `:recents` key.

calherries commented on src/metabase/api/activity.clj:
```diff
@@ -112,11 +116,39 @@
+  {context :string}
+  (let [context (if (str/blank? context)
+                  (throw (ex-info "Context required" {:context context}))
+                  (mapv keyword (str/split context #",")))]
```
To pass a list of values for a query parameter, we typically repeat the param name like `GET api/recents?context=view&context=collection`.

See this for an example, with the `filtering` parameter:
https://github.com/metabase/metabase/blob/5b01cf86d7bd4730140fdf8e6c438690de697be6/src/metabase/api/dashboard.clj#L1086-L1115

escherize replied:
Nice helper!

sloansparger committed:
refactors FE around new endpoint

sloansparger committed:
fixes for unit tests

escherize committed:
use ms/QueryVectorOf for context

escherize committed:
Merge branch 'recents-with-context' of github.com:metabase/metabase into recents-with-context

escherize committed:
fix models/recent_views tests

sloansparger committed:
use multiple query params instead of comma delimited value on FE

escherize committed:
ignore timestamp when deduping recents

escherize committed:
Merge branch 'recents-with-context' of github.com:metabase/metabase into recents-with-context

escherize committed:
review comments + test fixing

escherize committed:
update docstring

escherize committed:
fix api/activity_test s

escherize committed:
Merge branch 'master' into recents-with-context

escherize committed:
actually dedupe

escherize committed:
add test for deduping by context

escherize committed:
Merge branch 'recents-with-context' of github.com:metabase/metabase into recents-with-context

escherize committed:
e2e fix: shows up-to-date list of recently viewed items after another page is visited

escherize committed:
e2e fix: should undo the question replace action

escherize committed:
e2e fix: should replace a dashboard card question (metabase#36984)

escherize committed:
e2e fix: should preselect the most recently visited dashboard

escherize committed:
fix 6 more e2e tests

escherize committed:
Merge branch 'master' into recents-with-context

sloansparger committed:
fixes fe type check and unit failure

sloansparger committed:
renames unit test mocking function

iethree commented on frontend/src/metabase-types/api/activity.ts:
```diff
@@ -48,8 +59,24 @@ export interface RecentItemsResponse {
+export interface CreateRecentRequest {
+  model_id: number;
+  model: ActivityModel;
+  context: "selection";
```
I know this request will always be "selection" but can we define a type for activity context? `"view" | "selection"` ? This now comes back on RecentItem, right?

sloansparger replied:
there is a type for the GET request, those values are plural `"views" | "selections"`.

escherize replied:
Making sure we are all on the same page: the POST endpoint this interface is used for doesn't allow adding recent items with `context=view`, since we don't need it. Only "selection" is valid there at this time

escherize committed:
Merge branch 'master' into recents-with-context

rafpaf commented on e2e/test/scenarios/onboarding/command-palette.cy.spec.js:
```diff
@@ -108,12 +108,6 @@ describe("command palette", () => {
-      .should("have.attr", "aria-selected", "true");
-
-    pressPageDown();
-
```
just curious why this is getting removed here

rafpaf commented on frontend/src/metabase-types/api/activity.ts:
```diff
@@ -48,8 +59,24 @@ export interface RecentItemsResponse {
   recent_views: RecentItem[];
 }
 
+export type RecentContexts = "selections" | "views";
```
not `RecentsContexts`?

rafpaf commented on frontend/src/metabase/api/activity.ts:
```diff
@@ -1,22 +1,35 @@
+        const contextParam = [...context]
+          .sort()
+          .map(ctx => `context=${ctx}`)
+          .join("&");
```
nit, nonblocking, using `URLSearchParams` might be tidier and more explicit here

ranquild replied:
`api` supports params and just body; both will handle query string params for `GET`

escherize committed:
fix a flaky e2e test

rafpaf commented on frontend/src/metabase/browse/components/BrowseModels.tsx:
```diff
@@ -59,7 +59,7 @@ export const BrowseModels = () => {
 
-  const recentModelsResult = useListRecentItemsQuery(undefined, {
+  const recentModelsResult = useListRecentsQuery(undefined, {
     refetchOnMountOrArgChange: true,
```
since this is a request for recent views, we can add `context=["views"]` here. Or I can do that in a separate PR

rafpaf reviewed (approved):
FE looks good

escherize committed:
widen Item to accept str or kw where sensible

- allow strings or keywords for moderated_status, and authority_level

escherize committed:
Merge branch 'recents-with-context' of github.com:metabase/metabase into recents-with-context

escherize committed:
simplify impl + add test

escherize committed:
add view-log to events schema

escherize committed:
add collection context to view log

escherize committed:
fix the final 2 failing e2e tests

escherize committed:
click dashboard tab when the user has can-read? on recent entities
